File: python/033. Search in Rotated Sorted Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        left, right = 0, n-1
        while left <= right:
            mid = (left + right) / 2
            if nums[mid] == target: return mid
            if nums[mid] > nums[left]:
                if nums[left] <= target and target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            elif nums[mid] < nums[left]:
                if nums[mid] < target and target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                left += 1
        return -1

File: C++/033. Search in Rotated Sorted Array.cpp
class Solution {
private:
    int find(vector<int> &A, int target, int left, int right)
    {
        if (left > right) return -1;
        if (left == right) return (A[left] == target)? left:-1;
        int idx = -1;
        if (A[left] <= target && target <= A[right])
        {
            while(left <= right)
            {
                int mid = left + ((right - left) >> 1);
                if (A[mid] == target) return mid;
                else if (A[mid] < target)
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }
        }
        else
        {
            int mid  = left + ((right - left) >> 1);
            if (A[mid] == target) return mid;
            else
            {
                idx = find(A, target, left, mid-1);
                return (idx == -1)? find(A, target, mid+1, right) : idx;
            }
        }
        return idx;
    }
public:
    int search(vector<int> &A, int target) 
    {
        // write your code here
        int n = A.size();
        return find(A, target, 0, n-1);
    }
};
